### PR TITLE
feat(pie): Better pie label positioning

### LINKED ISSUE #2956: Pie chart tick labels overlap for smaller values

### LINKED ISSUE #2994: Pie Chart Visualization Issue
In our case, we have few pie charts where they are not loading when the legend text size is long. Could there be a way to manipulate the text size of the  legend.
causation``

### PR DESCRIPTION
**Pull request by jbrundage**
Fixes #2994
Fixes #2956

Signed-off-by: Jaman Brundage <jbrundage372@gmail.com>

### CODE REVIEW

File: packages/chart/src/Pie.ts [hunk 1]
```diff
@@ -270,8 +405,17 @@ Pie.prototype.mixin(Utility.SimpleSelectionMixin);
...
+    startAngle(_: number): this;
+    labelHeight(): number;
+    labelHeight(_: number): this;
 }
 Pie.prototype.publish("showSeriesPercentage", false, "boolean", "Append data series percentage next to label");
 Pie.prototype.publish("paletteID", "default", "set", "Color palette for this widget", Pie.prototype._palette.switch(), { tags: ["Basic", "Shared"] });
 Pie.prototype.publish("useClonedPalette", false, "boolean", "Enable or disable using a cloned palette", null, { tags: ["Intermediate", "Shared"] });
 Pie.prototype.publish("innerRadius", 0, "number", "Sets inner pie hole radius as a percentage of the radius of the pie chart", null, { tags: ["Basic"], range: { min: 0, step: 1, max: 100 } });
+Pie.prototype.publish("minOuterRadius", 20, "number", "Minimum outer radius (pixels)");
+Pie.prototype.publish("startAngle", -Math.PI, "number", "Starting angle of the first (and largest) wedge (radians)");
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Should be in degrees (for mortals)

Comment (jbrundage) [reply packages/chart/src/Pie.ts]: Done, and I added math.ts to util

File: tests/test-util/src/math.spec.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { degreesToRadians, radiansToDegrees } from "@hpcc-js/util";
+import { expect } from "chai";
+
+describe("math", function () {
+    it("degreesToRadians", function () {
+        expect(degreesToRadians(0)).to.equal(0);
+    });
+    it("radiansToDegrees", function () {
+        expect(radiansToDegrees(0)).to.equal(0);
+    });
```
Comment (GordonSmith) [inline tests/test-util/src/math.spec.ts]: Add more tests (all functions should be well covered) - try and focus on the boundary conditions (like 0, 2PI etc).

File: packages/util/src/math.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+export function normalizeDegrees(degrees: number, min: number = -180, max: number = 180): number {
+    return normalize(degrees, min, max);
+}
+
+export function normalize(value: number, min: number, max: number): number {
+    const spread = max - min;
+    const offsetValue = value - min;
+    return (offsetValue - (Math.floor(offsetValue / spread) * spread)) + min;
+}
```
Comment (GordonSmith) [inline packages/util/src/math.ts]: Add header comments (since we do that for util).

File: packages/util/src/index.ts [hunk 4]
```diff
@@ -7,6 +7,7 @@ export * from "./graph";
 export * from "./hashSum";
 export * from "./immutable";
 export * from "./logging";
+export * from "./math";
```
Comment (GordonSmith) [inline packages/util/src/index.ts]: Would "geom" be a better file name?

File: packages/chart/src/Pie.ts [hunk 5]
```diff
@@ -270,8 +393,17 @@ Pie.prototype.mixin(Utility.SimpleSelectionMixin);
...
+    startAngle(_: number): this;
+    labelHeight(): number;
+    labelHeight(_: number): this;
 }
 Pie.prototype.publish("showSeriesPercentage", false, "boolean", "Append data series percentage next to label");
 Pie.prototype.publish("paletteID", "default", "set", "Color palette for this widget", Pie.prototype._palette.switch(), { tags: ["Basic", "Shared"] });
 Pie.prototype.publish("useClonedPalette", false, "boolean", "Enable or disable using a cloned palette", null, { tags: ["Intermediate", "Shared"] });
 Pie.prototype.publish("innerRadius", 0, "number", "Sets inner pie hole radius as a percentage of the radius of the pie chart", null, { tags: ["Basic"], range: { min: 0, step: 1, max: 100 } });
+Pie.prototype.publish("minOuterRadius", 20, "number", "Minimum outer radius (pixels)");
+Pie.prototype.publish("startAngle", -180, "number", "Starting angle of the first (and largest) wedge (degrees)");
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: "0" should be the top, not the bottom...

File: packages/chart/src/Pie.ts [hunk 6]
```diff
@@ -187,30 +245,26 @@ export class Pie extends SVGWidget {
         text.exit()
             .remove();
 
-        const polyline = this._labels.selectAll("polyline").data(this.d3Pie(this.data()), d => this.getLabelText(d));
+        const polyline = this._labels.selectAll("polyline").data(this.d3Pie(this.data()), d => this.getLabelText(d, true));
 
         polyline.enter()
             .append("polyline")
             .merge(polyline)
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Add `.transition()` to help with the animation.

File: packages/chart/src/Pie.ts [hunk 7]
```diff
@@ -155,30 +201,42 @@ export class Pie extends SVGWidget {
...
+                });
+            });
+        this.adjustForOverlap();
+        mergedText
+            .style("font-size", this.labelHeight() + "px")
+            // .style("fill", function (d, i) {
+            //     const mid_angle = midAngle(d);
+            //     const quad = context.getQuadrant(mid_angle);
+            //     return context.getQuadrantColor(quad);
+            // })
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: remove commented code

File: packages/chart/src/Pie.ts [hunk 8]
```diff
@@ -155,30 +201,42 @@ export class Pie extends SVGWidget {
                 context.dblclick(context.rowToObj(d.data), context.columns()[1], context._selection.selected(this));
             })
             .merge(text)
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Add `.transition()` to help with animation.

File: demos/dermatology/tests/chartFactory.js [hunk 9]
```diff
@@ -412,10 +412,43 @@
...
+                            ["Randomized Algorithms", 41],
+                            ["Queueing Theory", 42],
+                        ])
                     );
                 });
+            },
+            many_small: function (callback) {
+                legacyRequire(["test/DataFactory", "src/chart/Pie"], function (DataFactory, Pie) {
+                    const pie = new Pie()
+                        .startAngle(0.3)
```
Comment (GordonSmith) [inline demos/dermatology/tests/chartFactory.js]: Update to use degrees

File: packages/chart/src/Pie.ts [hunk 10]
```diff
@@ -187,30 +245,26 @@ export class Pie extends SVGWidget {
...
+                const mid_angle = midAngle(d);
+                pos[0] = labelRadius * (context.isLeftSide(mid_angle) ? 1 : -1);
+                pos[1] = context._labelPositions[i].top;
+                return [pos1, pos2, pos];
             });
 
         polyline.exit()
             .remove();
 
+        this.centerOnLabels();
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Called twice for each `update`?

**Comment by jbrundage** [comment]
@GordonSmith These items have been resolved except for removal of the `calcSmallValueLabelHeight` function. It approximates the amount of additional vertical space needed for labels that are likely to cause overlap. There's a more accurate solution (calculating the exact amount of additional vertical space needed), but it will have edge cases when `startAngle` is involved. Would you like me to look into that accurate solution?

File: demos/dermatology/tests/chartFactory.js (hunk 9 above)
Comment (GordonSmith) [reply demos/dermatology/tests/chartFactory.js]: Still using radians???

File: demos/dermatology/tests/chartFactory.js [hunk 11]
```diff
@@ -412,10 +412,43 @@
...
+                        .columns(DataFactory.TwoD.subjects.columns)
+                        .data(Array(26).fill(1).map((n, i) => {
+                            return [Array(20).fill(String.fromCharCode(65 + i)).join(''), i + 1];
+                        }))
+                        ;
+                    setInterval(function () {
+                        pie
+                            .startAngle(pie.startAngle() + 0.3)
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply demos/dermatology/tests/chartFactory.js]: Stillusing radians???

File: demos/dermatology/tests/chartFactory.js [hunk 12]
```diff
@@ -412,10 +412,43 @@
...
+                            .render()
+                            ;
+                    }, 1500)
```
Comment (GordonSmith) [inline demos/dermatology/tests/chartFactory.js]: You can lower to 300 for nicer affect,

File: packages/chart/src/Pie.ts [hunk 13]
```diff
@@ -1,8 +1,11 @@
 import { I2DChart, ITooltip } from "@hpcc-js/api";
 import { InputField, SVGWidget, Utility } from "@hpcc-js/common";
+import { degreesToRadians, normalizeRadians } from "@hpcc-js/util";
 import { interpolate as d3Interpolate } from "d3-interpolate";
 import { select as d3Select } from "d3-selection";
 import { arc as d3Arc, pie as d3Pie } from "d3-shape";
+import { transition as d3Transition } from "d3-transition";
+d3Select.prototype.transition = d3Transition;
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: This does not look correct to me?

File: packages/chart/src/Pie.ts [hunk 14]
```diff
@@ -155,62 +203,55 @@ export class Pie extends SVGWidget {
...
+                    top: pos[1],
+                    bottom: pos[1] + context.labelHeight()
+                });
             });
-
-        function midAngle(d) {
-            return d.startAngle + (d.endAngle - d.startAngle) / 2;
-        }
+        this.adjustForOverlap();
+        mergedText
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Probably better to add transition here...

File: packages/chart/src/Pie.ts [hunk 15]
```diff
@@ -155,62 +203,55 @@ export class Pie extends SVGWidget {
...
-                return function (t) {
-                    const d2 = interpolate(t);
-                    const pos = context.d3LabelArc.centroid(d2);
-                    const pos1 = context.d3Arc.centroid(d2);
-                    const pos2 = [...pos];
-                    const mid_angle = midAngle(d2);
-                    pos[0] = labelRadius * (mid_angle < Math.PI && mid_angle > 0 ? 1 : -1);
-                    return [pos1, pos2, pos];
-                };
+            .transition()
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: By convention I typically put the transition on the same line as the previous call:  https://github.com/hpcc-systems/Visualization/blob/master/packages/chart/src/Pie.ts#L119

File: packages/chart/src/Pie.ts [hunk 16]
```diff
@@ -220,14 +261,100 @@ export class Pie extends SVGWidget {
...
     exit(_domNode, _element) {
         SVGWidget.prototype.exit.apply(this, arguments);
     }
 
     updateD3Pie() {
+        const startAngle = normalizeRadians(degreesToRadians(this.startAngle() - 180));
         this.d3Pie
             .padAngle(0.0025)
+            .startAngle(-Math.PI + startAngle)
+            .endAngle(Math.PI + startAngle)
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Can you remove the subtraction of 180 degrees and simply change the use of PI here?
(180 degrees === Math.PI in radians)

File: packages/util/src/math.ts [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * @param value value to be normalized
+ * @param min lower limit
+ * @param max upper limit
+ */
+export function normalize(value: number, min: number, max: number): number {
+    const spread = max - min;
+    const offsetValue = value - min;
+    return (offsetValue - (Math.floor(offsetValue / spread) * spread)) + min;
+}
```
Comment (GordonSmith) [inline packages/util/src/math.ts]: The headers blocks should have a `@returns XXX` for consistency

File: packages/chart/src/Pie.ts (hunk 13 above)
Comment (jbrundage) [reply packages/chart/src/Pie.ts]: Without it, the linter complains about transition not existing on `Selection`

Comment (GordonSmith) [reply packages/chart/src/Pie.ts]: Change to: 
```
import "d3-transition";
```

File: packages/chart/src/Pie.ts [hunk 18]
```diff
@@ -154,63 +202,53 @@ export class Pie extends SVGWidget {
...
-            .styleTween("text-anchor", function (d) {
-                this._current = this._current || d;
-                const interpolate = d3Interpolate(this._current, d);
-                this._current = interpolate(0);
-                return function (t) {
-                    const d2 = interpolate(t);
-                    const mid_angle = midAngle(d2);
-                    return mid_angle < Math.PI && mid_angle > 0 ? "start" : "end";
-                };
+            .merge(text).transition()
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Remove `transition()` from here as its already at Line 217

File: demos/dermatology/tests/chartFactory.js [hunk 19]
```diff
@@ -412,10 +412,43 @@
...
+                    const pie = new Pie()
+                        .startAngle(30)
+                        .columns(DataFactory.TwoD.subjects.columns)
+                        .data(Array(26).fill(1).map((n, i) => {
+                            return [Array(20).fill(String.fromCharCode(65 + i)).join(''), i + 1];
+                        }))
+                        ;
+                    setInterval(function () {
+                        pie
+                            .startAngle(pie.startAngle() + 30)
```
Comment (GordonSmith) [inline demos/dermatology/tests/chartFactory.js]: Change to "3"

File: demos/dermatology/tests/chartFactory.js (hunk 12 above)
Comment (GordonSmith) [reply demos/dermatology/tests/chartFactory.js]: Change to 300

File: packages/chart/src/Pie.ts [hunk 20]
```diff
@@ -21,6 +24,12 @@ export class Pie extends SVGWidget {
     d3Pie;
     d3Arc;
     d3LabelArc;
+    _labelPositions;
+    _smallValueLabelHeight;
+    _labelWidthLimit: number;
+    _quadIdxArr;
+    _minLabelTop = 0;
+    _maxLabelBottom = 0;
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: These should all be private?

File: packages/chart/src/Pie.ts [hunk 21]
```diff
@@ -127,21 +175,21 @@ export class Pie extends SVGWidget {
                     ;
             })
             ;
-
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Leave space in

File: packages/chart/src/Pie.ts [hunk 22]
```diff
@@ -127,21 +175,21 @@ export class Pie extends SVGWidget {
...
         //  Exit  ---
         arc.exit().transition()
             .style("opacity", 0)
             .remove()
             ;
-
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Leave space in

File: packages/chart/src/Pie.ts [hunk 23]
```diff
@@ -154,63 +202,53 @@ export class Pie extends SVGWidget {
...
 
         polyline.exit()
             .remove();
 
+        this.centerOnLabels();
+
+        function midAngle(d) {
+            return d.startAngle + (d.endAngle - d.startAngle) / 2;
+        }
         function arcTween(outerRadiusDelta, delay) {
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Add space between methods for consistency

File: packages/chart/src/Pie.ts [hunk 24]
```diff
@@ -220,14 +258,94 @@ export class Pie extends SVGWidget {
             };
         }
     }
-
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Add space between methods for consistency

File: packages/chart/src/Pie.ts [hunk 25]
```diff
@@ -220,14 +258,94 @@ export class Pie extends SVGWidget {
...
+    isLeftSide(midAngle) {
+        midAngle = normalizeRadians(midAngle);
+        const isLeft = midAngle > Math.PI * 2 ? midAngle : midAngle < Math.PI && midAngle > 0;
+        return isLeft;
+    }
+    getQuadrant(radians) {
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Add space between methods for consistency

File: packages/chart/src/Pie.ts [hunk 26]
```diff
@@ -220,14 +258,94 @@ export class Pie extends SVGWidget {
...
+        let quad = 0;
+        const rad = normalizeRadians(radians);
+        quad = rad <= Math.PI * 1.0 && rad >= Math.PI * 0.5 ? 3 : quad;
+        quad = rad <= Math.PI * 0.5 && rad >= Math.PI * 0.0 ? 2 : quad;
+        quad = rad <= Math.PI * 0.0 && rad >= Math.PI * -0.5 ? 1 : quad;
+        return quad;
+    }
+    centerOnLabels() {
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Add space between methods for consistency

File: packages/chart/src/Pie.ts [hunk 27]
```diff
@@ -220,14 +258,94 @@ export class Pie extends SVGWidget {
...
+            yShift = absTop - gY + (this.labelHeight() / 2);
+            yShift += heightDiff / 2;
+        }
+        const pos = this.pos();
+        this.pos({
+            y: pos.y + yShift,
+            x: pos.x
+        });
+    }
+    adjustForOverlap() {
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Add space between methods for consistency

File: packages/chart/src/Pie.ts [hunk 28]
```diff
@@ -220,14 +258,94 @@ export class Pie extends SVGWidget {
...
+                if (this._minLabelTop > this._labelPositions[n].top) {
+                    this._minLabelTop = this._labelPositions[n].top;
+                }
+                if (this._maxLabelBottom < this._labelPositions[n].bottom) {
+                    this._maxLabelBottom = this._labelPositions[n].bottom;
+                }
+            });
+        });
+    }
     exit(_domNode, _element) {
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: Add space between methods for consistency

File: demos/dermatology/tests/chartFactory.js (hunk 12 above)
Comment (jbrundage) [reply demos/dermatology/tests/chartFactory.js]: Nevermind, one of the transition calls was causing the labels to transition twice.

File: packages/chart/src/Pie.ts [hunk 29]
```diff
@@ -17,10 +19,15 @@ export class Pie extends SVGWidget {
 
     protected _totalValue: number;
 
-    labelWidgets;
-    d3Pie;
-    d3Arc;
-    d3LabelArc;
+    private d3Pie;
+    private d3Arc;
+    private d3LabelArc;
```
Comment (GordonSmith) [inline packages/chart/src/Pie.ts]: I suspect the d3Pie / Arc and Label should not be modified - I was referung to your new properties specific to the label layouts.

File: packages/chart/src/Pie.ts (hunk 21 above)
Comment (GordonSmith) [reply packages/chart/src/Pie.ts]: Bump

File: packages/chart/src/Pie.ts (hunk 22 above)
Comment (GordonSmith) [reply packages/chart/src/Pie.ts]: Bump

File: packages/chart/src/Pie.ts (hunk 23 above)
Comment (GordonSmith) [reply packages/chart/src/Pie.ts]: Bump
